executor/write.go

```diff
@@ -933,43 +977,183 @@ func getKeysNeedCheck(ctx sessionctx.Context, t table.Table, rows [][]types.Datu
 			var key []byte
 			var distinct bool
 			key, distinct, err = v.GenIndexKey(ctx.GetSessionVars().StmtCtx,
-				colVals, recordIDs[i], nil)
+				colVals, handles[i], nil)
 			if err != nil {
 				return nil, errors.Trace(err)
 			}
 			if !distinct {
 				continue
 			}
-			keysWithErr = append(keysWithErr, keyWithDupError{key, kv.ErrKeyExists.FastGen("Duplicate entry '%d' for key '%s'", recordIDs[i], v.Meta().Name)})
+			keysWithErr = append(keysWithErr, keyWithDupError{
+				false,
+				key,
+				kv.ErrKeyExists.FastGen("Duplicate entry '%d' for key '%s'", handles[i], v.Meta().Name),
+				handles[i],
+				newRowValue,
+			})
 		}
 		rowWithKeys = append(rowWithKeys, keysWithErr)
 	}
 	return rowWithKeys, nil
 }
 
-// batchMarkDupRows marks rows with duplicate errors as nil.
-// All duplicate rows were marked and appended as duplicate warnings
-// to the statement context in batch.
-func batchMarkDupRows(ctx sessionctx.Context, t table.Table, rows [][]types.Datum) ([][]types.Datum, error) {
+func batchGetInsertRows(ctx sessionctx.Context, t table.Table, newRows [][]types.Datum) ([][]keyWithDupError, map[string][]byte, error) {
```

Comment:
I can tell you change the name of this method. How about the behavior? Are they consistent? Some comment can resolve these confusion.